File: csharp/src/CodexWebMcp/Codex.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace CodexWebMcp;

public static class Codex
{
    public static async Task<string> RunAsync(string prompt, CancellationToken ct = default)
    {
        string tmpFile = Path.Combine(Path.GetTempPath(), $"codex-out-{Guid.NewGuid():N}.txt");
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = Config.CodexBin,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            // Use ArgumentList — NOT Arguments — for correct quoting on all platforms
            psi.ArgumentList.Add("exec");
            psi.ArgumentList.Add("--skip-git-repo-check");
            psi.ArgumentList.Add("--ephemeral");
            psi.ArgumentList.Add("--color");
            psi.ArgumentList.Add("never");
            psi.ArgumentList.Add("--sandbox");
            psi.ArgumentList.Add("read-only");
            psi.ArgumentList.Add("--output-last-message");
            psi.ArgumentList.Add(tmpFile);
            psi.ArgumentList.Add(prompt);

            using var proc = Process.Start(psi)
                ?? throw new InvalidOperationException("Process.Start returned null");

            // Close stdin so codex doesn't wait for any input
            try { proc.StandardInput.Close(); } catch { /* ignore */ }

            // Drain stdout & stderr concurrently — pipe buffer (~64KB) would block the child otherwise
            var stdoutTask = proc.StandardOutput.ReadToEndAsync();
            var stderrTask = proc.StandardError.ReadToEndAsync();

            using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            timeoutCts.CancelAfter(TimeSpan.FromSeconds(Config.CodexTimeoutSeconds));

            try
            {
                await proc.WaitForExitAsync(timeoutCts.Token);
            }
            catch (OperationCanceledException)
            {
                try { proc.Kill(entireProcessTree: true); } catch { /* ignore */ }
                return $"ERROR: codex timed out after {Config.CodexTimeoutSeconds}s";
            }

            // Discard stdout content but ensure the drain completes
            await stdoutTask;
            var stderr = await stderrTask;

            if (proc.ExitCode != 0)
                return $"ERROR: codex exited with code {proc.ExitCode}: {stderr.Trim()}";

            if (!File.Exists(tmpFile))
                return "ERROR: codex produced no output file";

            var content = await File.ReadAllTextAsync(tmpFile, ct);
            if (string.IsNullOrWhiteSpace(content))
                return "ERROR: codex output file empty";

            return content.TrimEnd();
        }
        catch (Exception ex)
        {
            return $"ERROR: codex invocation failed: {ex.Message}";
        }
        finally
        {
            try { if (File.Exists(tmpFile)) File.Delete(tmpFile); } catch { /* ignore */ }
        }
    }
}

File: java/src/main/java/codexwebmcp/Codex.java
package codexwebmcp;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

public final class Codex {
    private Codex() {}

    public static String run(String prompt) {
        Path tmpPath = Path.of(System.getProperty("java.io.tmpdir"),
            "codex-out-" + UUID.randomUUID().toString().replace("-", "") + ".txt");

        List<String> args = new ArrayList<>();
        args.add(Config.codexBin());
        args.add("exec");
        args.add("--skip-git-repo-check");
        args.add("--ephemeral");
        args.add("--color"); args.add("never");
        args.add("--sandbox"); args.add("read-only");
        args.add("--output-last-message"); args.add(tmpPath.toString());
        args.add(prompt);

        ProcessBuilder pb = new ProcessBuilder(args);
        pb.redirectInput(ProcessBuilder.Redirect.from(devNull()));
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        pb.redirectError(ProcessBuilder.Redirect.PIPE);

        Process process;
        try {
            process = pb.start();
        } catch (IOException e) {
            cleanup(tmpPath);
            return "ERROR: codex invocation failed: " + e.getMessage();
        }

        StringBuilder stderr = new StringBuilder();
        Thread stderrThread = new Thread(() -> {
            try (var is = process.getErrorStream()) {
                byte[] buf = new byte[8192];
                int n;
                while ((n = is.read(buf)) > 0) {
                    stderr.append(new String(buf, 0, n, StandardCharsets.UTF_8));
                }
            } catch (IOException ignored) {
                // best effort
            }
        }, "codex-stderr-drain");
        stderrThread.setDaemon(true);
        stderrThread.start();

        try {
            boolean exited = process.waitFor(Config.CODEX_TIMEOUT_SECONDS, TimeUnit.SECONDS);
            if (!exited) {
                process.destroyForcibly();
                return "ERROR: codex timed out after 180s";
            }
            stderrThread.join(2000);

            int code = process.exitValue();
            if (code != 0) {
                return "ERROR: codex exited with code " + code + ": " + stderr.toString().trim();
            }
            if (!Files.exists(tmpPath)) {
                return "ERROR: codex produced no output file";
            }
            if (Files.size(tmpPath) == 0L) {
                return "ERROR: codex output file empty";
            }
            String data = Files.readString(tmpPath);
            String trimmed = data.replaceAll("[ \\t\\r\\n]+$", "");
            if (trimmed.isEmpty()) return "ERROR: codex output file empty";
            return trimmed;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            process.destroyForcibly();
            return "ERROR: codex interrupted";
        } catch (IOException e) {
            return "ERROR: failed to read codex output: " + e.getMessage();
        } finally {
            cleanup(tmpPath);
        }
    }

    private static File devNull() {
        String os = System.getProperty("os.name", "").toLowerCase();
        return new File(os.contains("win") ? "NUL" : "/dev/null");
    }

    private static void cleanup(Path p) {
        try {
            Files.deleteIfExists(p);
        } catch (IOException ignored) {
            // best effort
        }
    }
}
